let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}

chunfengyao opened the pull request:
let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}. will close #92252

*change the output directory of native plugin's from `$HOME/.pub-cache/${plugin_path}` to `${flutter_module_path}/.android/plugins_build_output/${androidPlugin.name}`. Because the directory userd before was a shared directory,it will cause issue #92252 when run multiple task simultaneously.*

*List which issues are fixed by this PR. You must list at least one issue.*
#92252

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

chunfengyao committed:
closed #92252

let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}.

chunfengyao commented:
@christopherfujino 
Can you find developer to review this pull request for [this issue](https://github.com/flutter/flutter/issues/92252).
Thanks.

christopherfujino commented:
cc @stuartmorgan does this look like the right change?

stuartmorgan commented:
I'm not familiar with the internal build setup of Android plugins; @blasten would be a better reviewer.

blasten reviewed (changes requested):
This needs to be fixed in the template file. The fix is:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/android/host_app_common/app.tmpl/build.gradle.tmpl#L33

should have a section similar to:
https://github.com/flutter/flutter/blob/ad3fe21cbcf508a32ae6aa249152c557493726a1/packages/flutter_tools/templates/app_shared/android-java.tmpl/build.gradle#L20-L22

chunfengyao committed:
changes for PR-94645.

change the output directory to "${rootProject.buildDir}/${androidPlugin.name}"

chunfengyao commented:
> This needs to be fixed in the template file. The fix is:
> 
> https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/android/host_app_common/app.tmpl/build.gradle.tmpl#L33
> 
> should have a section similar to:
> 
> https://github.com/flutter/flutter/blob/ad3fe21cbcf508a32ae6aa249152c557493726a1/packages/flutter_tools/templates/app_shared/android-java.tmpl/build.gradle#L20-L22

I have changed the plugin's output directory to `${rootProject.buildDir}/${pluginName}` by [changes for PR-94645.](https://github.com/flutter/flutter/pull/94645/commits/2116be9789b9a8c2597f68eba5f73d7695691f1d), you can review this pr again.
Thanks @blasten

blasten commented:
can we add the subproject logic below this line https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/android/host_app_common/app.tmpl/build.gradle.tmpl#L33?

The reason is that the file you are changing is also used for other case where this change may or may not make sense.

chunfengyao commented:
> can we add the subproject logic below this line https://github.com/flutter/flutter/blob/master/packages/flutter_tools/templates/module/android/host_app_common/app.tmpl/build.gradle.tmpl#L33?
> 
> The reason is that the file you are changing is also used for other case where this change may or may not make sense.

In some case, the host gradle project may dependent the plugin's native module directly. and the templete may not work for this case.

such as:
We add `flutter_aliplayer` to flutter project's pubspec.yaml. and sync the host project's gradle file. and then we will see a native module named :`flutter_aliplayer` appear in module list, and then we can let host project dependent this module(sometime, this can aviod some .so library conflict).

chunfengyao commented:
@blasten 
Is the `${flutterModule}/build/host/${pluginName}` was a better directory I could use for this?

chunfengyao committed:
changes for PR-94645.

let the plugin's output generated in `flutterModule/build/host/${androidPlugin.name}`.

chunfengyao commented:
Hi~ @blasten 
The forked [repo](https://github.com/chunfengyao/flutter/tree/chunfengyao-patch-for-issue-92252) has add access for you. please have a look at this pull request. 
Thanks very much.

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -20,15 +21,28 @@ if (pluginsFile.exists()) {
     }
 }
 
+String flutterModulePath = project(':flutter').projectDir.parentFile.parentFile.getAbsolutePath()
```
use `moduleProjectRoot` since it's defined above

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -20,15 +21,28 @@ if (pluginsFile.exists()) {
         def pluginDirectory = new File(androidPlugin.path, 'android')
         assert pluginDirectory.exists()
         include ":${androidPlugin.name}"
         project(":${androidPlugin.name}").projectDir = pluginDirectory
```
why not setting `buildDir` in the plugin project right here?

chunfengyao replied:
I have tried to set buildDir here, but It's not work.
The plugin is added to host project by `include ":${androidPlugin.name}"`. 
and then the method: `org.gradle.initialization.DefaultProjectDescriptor#buildFile` in `gradle-core-6.1.1.jar` provide the buildDir location.
The build dir location provided by `org.gradle.initialization.DefaultProjectDescriptor#buildFile` is `getProjectDir + buildFileName` by default. 
And gradle only provide the setter for `buildFileName`. 
so I think It'll be difficult to modify the build dir by `org.gradle.initialization.DefaultProjectDescriptor`. the `projectsLoaded` code block will be a easier place.

chunfengyao committed:
use `moduleProjectRoot`.

use `moduleProjectRoot` rather than define another one.

skia-gold commented:
Gold has detected about 6 new digest(s) on patchset 4.
View them at https://flutter-gold.skia.org/cl/github/94645

christopherfujino commented:
@blasten can you take a look at this PR?

chunfengyao commented:
> @blasten friendly ping @chunfengyao to get the `customer_testing-linux` CI check to pass, you will have to rebase (not merge) upstream master.

thanks for tips, I rebased the branch with latest master.

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +37,17 @@ if (pluginsFile.exists()) {
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
+        //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
+        p.subprojects { sp ->
```
Just to clarify, Gradle subprojects don't equal Flutter plugins always. 
Flutter plugins is a subset of Gradle subprojects.

If you want to process the plugins only, move the logic below to line 34.  See `project(":${androidPlugin.name}")`, that is the subproject you are looking for.

chunfengyao replied:
@blasten Thanks for your advise. I have tried it just moment. But I cannot modify the build directory use `project(":${androidPlugin.name}")`. `project(":${androidPlugin.name}")` returns the class `org.gradle.initialization.DefaultProjectDescriptor`, It's only allowed me to modify "source dir" ,"(sub)project name" and "(sub)project build script file" . you can see [this doc](https://docs.gradle.org/current/javadoc/org/gradle/api/initialization/ProjectDescriptor.html). 
and `gradle.getGradle().projectsLoaded` returns the class 'org.gradle.api.internal.project.DefaultProject_Decorated' ,It is allowed to modify the build directory.

blasten replied:
In this case can we add comments?

blasten reviewed (changes requested):
needs test

blasten commented:
For the test, https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/android_plugin_compilesdkversion_mismatch_test.dart is an example of how Gradle intergrations are currently tested.

chunfengyao commented:
> For the test, https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/android_plugin_compilesdkversion_mismatch_test.dart is an example of how Gradle intergrations are currently tested.

@blasten I have ran the test , and got no error.
```
/Volumes/Storage/AS_WS/flutter_issue_handle/bin/cache/dart-sdk/bin/dart run test -r json /Volumes/Storage/AS_WS/flutter_issue_handle/packages/flutter_tools/test/integration.shard/android_plugin_compilesdkversion_mismatch_test.dart
Testing started at 11:57 ...

Process finished with exit code 0
```

chunfengyao committed:
add necessary code comments.

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +38,19 @@ if (pluginsFile.exists()) {
     g.rootProject.beforeEvaluate { p ->
+        //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
+        p.subprojects { sp ->
+            if (pluginsNames.contains(sp.name.replace(":", ""))){
```
does `name` contain `:`? in my local test it doesn't.

blasten replied:
also can this just be `object.plugins.android`? that is a list already.

chunfengyao replied:
maybe someone else modified the `name` on `master` with another pr. I checked my local project , the `name` does not contained the `:` now.

blasten replied:
ok. it might have been a recent Gradle version bump. Let's keep the logic that strips `:` out in the variable specified below.

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -20,6 +21,8 @@ if (pluginsFile.exists()) {
     object.plugins.android.each { androidPlugin ->
         assert androidPlugin.name instanceof String
         assert androidPlugin.path instanceof String
+        // save plugin's name for modifying it's buildDir location. see issue #92252 for detail.
+        pluginsNames.add(androidPlugin.name)
```
```suggestion
```

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -11,6 +11,7 @@ def moduleProjectRoot = project(':flutter').projectDir.parentFile.parentFile
 
 // Note: if this logic is changed, also change the logic in app_plugin_loader.gradle.
 def pluginsFile = new File(moduleProjectRoot, '.flutter-plugins-dependencies')
+List<String> pluginsNames = new ArrayList<>()
```
```suggestion
```

blasten replied:
this isn't resolved yet. Did you try using the existing list?

chunfengyao replied:
It was resolved by the last commit. It has been changed to use `object.plugins.android.name`.
[changed file](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/gradle/module_plugin_loader.gradle)

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +38,19 @@ if (pluginsFile.exists()) {
     g.rootProject.beforeEvaluate { p ->
+        //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
+        p.subprojects { sp ->
+            if (pluginsNames.contains(sp.name.replace(":", ""))){
```
```suggestion
            if (object.plugins.android.contains(sp.name)){
```

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +38,19 @@ if (pluginsFile.exists()) {
+        //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
+        p.subprojects { sp ->
+            if (pluginsNames.contains(sp.name.replace(":", ""))){
+                File androidPluginBuildOutputDir = new File(moduleProjectRoot.getAbsolutePath() + File.separator + "build" + File.separator + "host" + File.separator + sp.name.replace(":", ""));
```
```suggestion
                File buildDir = new File(moduleProjectRoot.getAbsolutePath() + File.separator + "build" + File.separator + "host" + File.separator + sanitizedPluginName);
```

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +38,19 @@ if (pluginsFile.exists()) {
     g.rootProject.beforeEvaluate { p ->
+        //let the plugin's output generated in flutterModule/build/host/${androidPlugin.name}.
+        p.subprojects { sp ->
+            if (pluginsNames.contains(sp.name.replace(":", ""))){
```
```suggestion
            def sanitizedPluginName = sp.name.replace(":", "")
            if (object.plugins.android.contains(sanitizedPluginName)) {
```

chunfengyao replied:
where is the `object` defined at `object.plugins.android.contains(sanitizedPluginName)`?

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +38,19 @@ if (pluginsFile.exists()) {
+        p.subprojects { sp ->
+            if (pluginsNames.contains(sp.name.replace(":", ""))){
+                File androidPluginBuildOutputDir = new File(moduleProjectRoot.getAbsolutePath() + File.separator + "build" + File.separator + "host" + File.separator + sp.name.replace(":", ""));
+                if (!androidPluginBuildOutputDir.exists()){
+                    //make dir if not exist.
+                    androidPluginBuildOutputDir.mkdirs()
+                }
+                // only the project object here can modify the `buildDir`.
+                // {@see https://github.com/flutter/flutter/pull/94645/files/7c77d3653709ccd1e2b45268b209f00f36d8470d#r819224682}
+                sp.buildDir = androidPluginBuildOutputDir
```
```suggestion
                // Create the directory recursively if it doesn't exist.
                if (!buildDir.exists()) {
                    buildDir.mkdirs()
                }
                sp.buildDir = buildDir
```

blasten commented:
please try the changes out. Also, we would need a test for this.

You should be able to run the sample test if you `cd` into `packages/flutter_tool`, then run `flutter test test/integration.shard/android_plugin_compilesdkversion_mismatch_test.dart`. See:

```
$ flutter test test/integration.shard/android_plugin_compilesdkversion_mismatch_test.dart
Running "flutter pub get" in flutter_tools...                    1,753ms
00:50 +1: All tests passed!                                                                                                                         
```

chunfengyao committed:
Update module_plugin_loader.gradle

Get plugin's name from `.flutter-plugins-dependencies`.

chunfengyao committed:
update code comment.

code comment.

chunfengyao commented:
Should i merge the lastest master branch to pass the checks?

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -9,13 +9,16 @@ import groovy.json.JsonSlurper
     assert object instanceof Map
     assert object.plugins instanceof Map
     assert object.plugins.android instanceof List
+    androidPlugins = object.plugins.android
```
I would move `object` definition to line 12 instead.

chunfengyao replied:
The `object `was generated by `pluginsFile` which defined at line 15. Isn't the `object` must defined below `pluginsFile`?

blasten replied:
you can add `def object = null` to line 12. Then remove `def` from line 17. then use object in line 43, and line 45

chunfengyao replied:
ok.

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -33,8 +36,25 @@ if (pluginsFile.exists()) {
+                        //make dir if not exist.
+                        androidPluginBuildOutputDir.mkdirs()
+                    }
+                    //let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}.
```
```suggestion
                    //let the plugin's output generated in flutter/.android/plugins_build_output/${androidPlugin.name}.
```

this comment does't add any new information

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -33,8 +36,25 @@ if (pluginsFile.exists()) {
+                if (androidPlugins.name.contains(sanitizedPluginName)){
+                    File androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator + "plugins_build_output" + File.separator + sanitizedPluginName);
+                    if (!androidPluginBuildOutputDir.exists()){
+                        //make dir if not exist.
```
```suggestion
```
This comment doesn't add any new information

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -33,8 +36,25 @@ if (pluginsFile.exists()) {
+            //only flutter submodules.
+            if (androidPlugins != null) {
+                String sanitizedPluginName = sp.name.replace(":", "");
+                //matching submodule, which one's name on the `pluginsFile`.
```
```suggestion
```

This comment doesn't add any new information

chunfengyao committed:
remove unnecessary code comments.

chunfengyao committed:
move `module_plugin_loader.gradle`'s `object` define to line 12.

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -33,8 +35,20 @@ if (pluginsFile.exists()) {
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
+        p.subprojects { sp ->
+            if (object != null && object.plugins.android.name.contains(sanitizedPluginName)) {
```
also check for `object.plugins != null && object.plugins.android != null`

chunfengyao replied:
ok.

chunfengyao committed:
null check.

blasten commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -9,10 +9,12 @@ import groovy.json.JsonSlurper
 def moduleProjectRoot = project(':flutter').projectDir.parentFile.parentFile
 
+def object = null;
+String flutterModulePath = project(':flutter').projectDir.parentFile.getAbsolutePath()
```
```suggestion
```

stuartmorgan commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +37,18 @@ if (pluginsFile.exists()) {
+            def sanitizedPluginName = sp.name.replace(":", "");
+            if (object != null && object.plugins != null && object.plugins.android != null
+                    && object.plugins.android.name.contains(sanitizedPluginName)) {
+                def androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator
```
Why is the file separator being added twice here?

chunfengyao replied:
This's a mistake. I fixed it just now. thanks.

stuartmorgan commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +37,18 @@ if (pluginsFile.exists()) {
+                    && object.plugins.android.name.contains(sanitizedPluginName)) {
+                def androidPluginBuildOutputDir = new File(flutterModulePath + File.separator + File.separator
+                        + "plugins_build_output" + File.separator + sanitizedPluginName);
+                if (!androidPluginBuildOutputDir.exists()){
```
The existing code in this file appears to have a space before `{` in conditionals; this should match.

stuartmorgan-g reviewed (changes requested):
Where is the test for this change? This seems like a subtle behavior that would be easy to regress without a test.

chunfengyao committed:
check if the output file is generated at expected location.

stuartmorgan-g reviewed (changes requested):
My question about tests hasn't been resolved; this is still missing tests.

chunfengyao commented:
> My question about tests hasn't been resolved; this is still missing tests.

@stuartmorgan  Is that mean that I should wrote a *_test.dart file to test this pr's changes? or just run the presented test file like [this](https://github.com/flutter/flutter/pull/94645#issuecomment-1068707684)? 
And if need to write a *_test.dart file,what should I do? [The wiki](https://github.com/flutter/flutter/wiki/Running-and-writing-tests) has no doc about how to add a test for flutter_tools(gradle). 
Thanks.

stuartmorgan commented:
> > My question about tests hasn't been resolved; this is still missing tests.
> 
> @stuartmorgan Is that mean that I should wrote a *_test.dart file to test this pr's changes? or just run the presented test file like [this](https://github.com/flutter/flutter/pull/94645#issuecomment-1068707684)? And if need to write a *_test.dart file,what should I do? [The wiki](https://github.com/flutter/flutter/wiki/Running-and-writing-tests) has no doc about how to add a test for flutter_tools(gradle). Thanks.

You need to write a test for this change. The comments from @blasten above were pointing you to example tests that you can model a new test on, not telling you to run those tests. (Existing tests are automatically run by our CI.)

chunfengyao commented:
> > > My question about tests hasn't been resolved; this is still missing tests.
> > 
> > 
> > @stuartmorgan Is that mean that I should wrote a *_test.dart file to test this pr's changes? or just run the presented test file like [this](https://github.com/flutter/flutter/pull/94645#issuecomment-1068707684)? And if need to write a *_test.dart file,what should I do? [The wiki](https://github.com/flutter/flutter/wiki/Running-and-writing-tests) has no doc about how to add a test for flutter_tools(gradle). Thanks.
> 
> You need to write a test for this change. The comments from @blasten above were pointing you to example tests that you can model a new test on, not telling you to run those tests. (Existing tests are automatically run by our CI.)

[The issue](https://github.com/flutter/flutter/issues/92252) methioned by this pr only can be reproduced by running multiple build task simultaneously. And [the issue](https://github.com/flutter/flutter/issues/92252) was appeared Occasionally. 
Should I add a dependence in pubspec.yaml like [this test](https://github.com/flutter/flutter/blob/master/dev/devicelab/bin/tasks/plugin_dependencies_test.dart) to test this pr? or [the test commited just now]([packages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart](https://github.com/flutter/flutter/pull/94645/commits/668b85cb31046bf980c42f0ce3ee9fe08478f446#diff-eb64c2de77732f6ecce73d9f5dd62d22b15d7f25f142949c44a8cf69b3ba760b)) is enogh? @stuartmorgan

stuartmorgan commented:
> [The issue](https://github.com/flutter/flutter/issues/92252) methioned by this pr only can be reproduced by running multiple build task simultaneously. And [the issue](https://github.com/flutter/flutter/issues/92252) was appeared Occasionally.

A test does not necessarily need to reproduce the original issue, it just has to make sure it didn't regress. In this case you are moving the build output from a shared location to a location within the project; having an output directory within the project is a deterministic behavior that you should be able to asset.

> I added a simple test file: [android_plugin_multiple_build_tasks_test.dart](https://github.com/flutter/flutter/pull/94645/%5Bpackages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart%5D(https://github.com/flutter/flutter/pull/94645/commits/668b85cb31046bf980c42f0ce3ee9fe08478f446#diff-eb64c2de77732f6ecce73d9f5dd62d22b15d7f25f142949c44a8cf69b3ba760b)).

A test that would not reliably fail if the issue regressed is not one that we would check in. It wouldn't actually prevent regressions, and best-case would add flake by failing rarely (which is something we explicitly don't want).

chunfengyao commented:
@stuartmorgan Ok, I finished the test sample. Now , the test [android_plugin_multiple_build_tasks_test.dart](https://github.com/flutter/flutter/pull/94645/%5Bpackages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart%5D(https://github.com/flutter/flutter/pull/94645/commits/668b85cb31046bf980c42f0ce3ee9fe08478f446#diff-eb64c2de77732f6ecce73d9f5dd62d22b15d7f25f142949c44a8cf69b3ba760b)) will failed when the issue regressed.

stuartmorgan commented:
> Now , the test [android_plugin_multiple_build_tasks_test.dart](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart) will failed when the issue regressed.

As far as I can tell you are just trying to recreate the race condition in the test, which is the opposite of what I suggested above, and has the problems that I specifically said we don't want in a test. Am I misunderstanding what the test is doing?

I would not expect you to need to run multiple builds to validate that the output location is as expected.

stuartmorgan commented:
> If the pipline ran and passed this test , It can make sure the issue was not regressed.

What line in the test will deterministically fail without the other changes in the PR?

chunfengyao commented:
> > If the pipline ran and passed this test , It can make sure the issue was not regressed.
> 
> What line in the test will deterministically fail without the other changes in the PR?

[Line 132](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart#L132) may failed if `plugin_test` and `plugin_bg_test` runned simultaneously like the `android_plugin_multiple_build_tasks_test.dart` said. 
The two build processes will all write plugin's output file to `${pub-cache}/image_picker/build/` if without this PR. and when the `plugin_test` start to write plugin's output file to `${pub-cache}/image_picker/build/` (this operation will override the existed file which generated by `plugin_bg_test` which ran first) , `plugin_bg_test` may have begin to cleanning the `${pub-cache}/image_picker/build/`, 
So the `plugin_test` will got an gradle exception : the file I generated just now was disappear (File not found exception) like [the issue](https://github.com/flutter/flutter/issues/92252#issuecomment-950484958) said ,cause It has been deleted by `plugin_bg_test`'s clean task. 
then the [Line 132](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart#L132) may failed.

stuartmorgan commented:
> > What line in the test will deterministically fail without the other changes in the PR?
> 
> @stuartmorgan [Line 132](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart#L132) may failed if `plugin_test` and `plugin_bg_test` runned simultaneously

"may fail" and "will deterministically fail" are very different things. Please re-read my comments above [here](https://github.com/flutter/flutter/pull/94645#issuecomment-1103850968) and [here](https://github.com/flutter/flutter/pull/94645#issuecomment-1105228598) about the problems with this approach and my suggested alternative.

chunfengyao commented:
> > > What line in the test will deterministically fail without the other changes in the PR?
> > 
> > 
> > @stuartmorgan [Line 132](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_multiple_build_tasks_test.dart#L132) may failed if `plugin_test` and `plugin_bg_test` runned simultaneously
> 
> "may fail" and "will deterministically fail" are very different things. Please re-read my comments above [here](https://github.com/flutter/flutter/pull/94645#issuecomment-1103850968) and [here](https://github.com/flutter/flutter/pull/94645#issuecomment-1105228598) about the problems with this approach and my suggested alternative.

It's hard to make a test example for this issue , because , It's difficult to monitor the changes of generated cache file. 
Maybe It's just a normal change? It seems like that It's just a config(the output path) changed?

so, does It mean there's no errors once the apk was sucessfully built([this test sample did the same thing](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/android_plugin_compilesdkversion_mismatch_test.dart)).

stuartmorgan commented:
Can you elaborate on why it's difficult to write a test that verifies that build output was created in an expected location? I'm not clear on where the issue is.

chunfengyao commented:
> Can you elaborate on why it's difficult to write a test that verifies that build output was created in an expected location? I'm not clear on where the issue is.
> 
> (I'm also not sure what you mean by "just a normal change")

Did you mean that the test file just need to verify that the output file was generated at the expect place? 
It seems that this test example does too little. (the purpose of our project's test example is prevent issue regress.)

chunfengyao commented:
I have create [an other test example](https://github.com/flutter/flutter/pull/94645/commits/fe5ded6a528ceb77e42c6c96032fb9245e6c852b).

stuartmorgan commented:
> > Can you elaborate on why it's difficult to write a test that verifies that build output was created in an expected location? I'm not clear on where the issue is.
> > (I'm also not sure what you mean by "just a normal change")
> 
> Did you mean that the test file just need to verify that the output file was generated at the expect place?

Yes. As I said initially: "having an output directory within the project is a deterministic behavior that you should be able to assert."

> It seems that this test example does too little.

What other behavioral changes is your PR making that it doesn't test?

> (the purpose of our project's test example is prevent issue regress.)

If having the build output be in the new location instead of a shared location fixes the issue, how would the issue regress without breaking a test checking that it is using the new location?

chunfengyao commented:
> 

Thank you very much. Is there anything else should I do for the next?

chunfengyao commented:
> 

@stuartmorgan Thank you very much. Is there anything else should I do for the next?

christopherfujino commented:
@chunfengyao looks like you still have CI failures

christopherfujino commented:
@chunfengyao looks like you still have CI failures

chunfengyao committed:
Fixed # remove redundant definition caused by the conflicts when rebase.

chunfengyao committed:
Fixed # unknown definition.

chunfengyao commented:
> @chunfengyao looks like you still have CI failures

@christopherfujino Thanks, It seems that errors were fixed now.

chunfengyao commented:
Is there anywhere I can find the detail about the failure of [google_maps_flutter_web](https://ci.chromium.org/ui/p/flutter/builders/try/Linux%20flutter_plugins/30655/overview) which reported by `Linux flutter_plugins`, so that I can fix it?
I can find the exit code of test command in luci only.

stuartmorgan commented:
It's in the `stdout` link for the failing step. That was an out-of-band failure unrelated to your PR though; you just need to merge in the latest `master`.

chunfengyao commented:
> It's in the `stdout` link for the failing step. That was an out-of-band failure unrelated to your PR though; you just need to merge in the latest `master`.

@stuartmorgan Thanks , All CI checks were passed after rebased to the latest master. 
The changes about tests that you requested [here](https://github.com/flutter/flutter/pull/94645#pullrequestreview-945853845) seems have been added, could you review this PR again and mark the requested change resolved.

chunfengyao committed:
Fixed # recreate the losted test file which losted when handle merge conflict.

chunfengyao commented:
Sorry for the losted test file, I created a new test sample. could you have a look at the [new test sample](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart)? and review this PR again. Thanks. @stuartmorgan

chunfengyao commented:
Sorry for the losted test file, I created a new test sample. could you have a look at the [new test sample](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart)? and review this PR again. Thanks. @stuartmorgan

chunfengyao committed:
Fixed # recreate the losted test file which losted when handle merge conflict.

christopherfujino commented:
> Sorry for the losted test file, I created a new test sample. could you have a look at the [new test sample](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart)? and review this PR again. Thanks. @stuartmorgan

You still have failing pre-submit checks, could you please address?

chunfengyao committed:
Update android_plugin_new_output_dir_test.dart

Fixed # recreate the losted test file which losted when handle merge conflict.

chunfengyao commented:
> > Sorry for the losted test file, I created a new test sample. could you have a look at the [new test sample](https://github.com/chunfengyao/flutter/blob/chunfengyao-patch-for-issue-92252/packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart)? and review this PR again. Thanks. @stuartmorgan
> 
> You still have failing pre-submit checks, could you please address?

@christopherfujino Thanks, I fixed the warning thrown by code lint , and all checks have passed now.

chunfengyao commented:
> > It's in the `stdout` link for the failing step. That was an out-of-band failure unrelated to your PR though; you just need to merge in the latest `master`.
> 
> @stuartmorgan Thanks , All CI checks were passed after rebased to the latest master. The changes about tests that you requested [here](https://github.com/flutter/flutter/pull/94645#pullrequestreview-945853845) seems have been added, could you review this PR again and mark the requested changes was resolved?

I cannot mark the changes you requested resolved. May be It is caused by the outdated of the changes requested? could you help this PR? Thanks. @stuartmorgan

chunfengyao commented:
can you help this PR again? from the result of CI check, I clicked `see review`, and then click `view changes`, but the origin change request has gone. The file which the changes requested on has been deleted by force-push caused by rebase to latest master. cause of the reason above, can you review this PR again? Thanks. @stuartmorgan

stuartmorgan commented on packages/flutter_tools/gradle/module_plugin_loader.gradle:
```diff
@@ -35,6 +37,18 @@ if (pluginsFile.exists()) {
 gradle.getGradle().projectsLoaded { g ->
     g.rootProject.beforeEvaluate { p ->
+        p.subprojects { sp ->
+            String sanitizedPluginName = sp.name.replace(":", "");
```
What do the input names look like here? This transformation looks like it could create collisions between plugins.

chunfengyao replied:
I changed the input name to `subproject`. Maybe It is better here.

stuartmorgan replied:
Changing the variable name doesn't address my question, which was about the values. What do the values here look like?

This transformation would cause "a:bc" and "ab:c" to have that same output directory, which would break things. I'm trying to understand what the possible inputs are to know if that's a problem.

chunfengyao replied:
The reason why the transformation was ben done here is that the subproject's name always looks like `:abcde` or `ab:c` , the name `abcde` should contained in `object.plugins.android.name` and the name `abc` will not, after we removed the `:`.

`object.plugins.android.name` was generated by `.flutter-plugins-dependencies`, It contains all the flutter plugin's native dependencies which will inject to host project.

chunfengyao replied:
So ,even if the host project contain a dependency named "ab:c", `object.plugins.android.name.contains(pluginName)` will exclude it.

stuartmorgan replied:
> The reason why the transformation was ben done here is that the subproject's name always looks like `:abcde` or `ab:c`

What are actual examples of plugin subproject names? Are they like the first example you gave, or the second?

> the name `abcde` should contained in `object.plugins.android.name` and the name `abc` will not, after we removed the `:`.

If `abcde` is in the set, then (per my comment above) both `:abcde` and `ab:cde` would pass, and would attempt to use the same build directory, which would be bad. Is there something I'm missing that would prevent that from happening? If not, that's a bug in the logic here.

GaryQian replied:
I agree that this string transformation here is likely to cause naming conflicts. It seems like it would be straightforward for two similarly named packages where one name was a substring of another or have different colon positioning to both pass this check and conflict.

GaryQian replied:
Would it be simpler to just treat the colon as a directory separator and have the files in a subdirectory, for example `ab:c` would map to `<blah>/ab/c`. Since the goal is to create unique separate build directories, removing identifying characters like ':' seems to be counterproductive.

chunfengyao replied:
With the latest master It seems that there's no `:` before project's name.
I removed It and the test file also returned 0.

chunfengyao replied:
@stuartmorgan Re-request review manually.

stuartmorgan commented:
@GaryQian should review this as well.

chunfengyao committed:
Fixed # change input name or it will create collisions.

chunfengyao committed:
Fixed # wrong assert info.

chunfengyao committed:
Fixed # use specific version of image_picker.

chunfengyao commented on packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart:
```diff
@@ -0,0 +1,73 @@
+    tryToDelete(tempDir);
+  });
+
+  test("error logged when plugin's output dir was flutter_project/", () async {
```
Yes, the assert notice was mean the opposite. I have corrected it now.

chunfengyao commented on packages/flutter_tools/test/integration.shard/android_plugin_new_output_dir_test.dart:
```diff
@@ -0,0 +1,73 @@
+
+    // Check outputDir existed
+    final Directory outputDir = fileSystem.directory(fileSystem.path.join(
+        projectPath, '.android', 'plugins_build_output', 'image_picker_android'
```
Thanks , I changed the `any` to a specific version number.

chunfengyao commented:
> @GaryQian should review this as well.

I clicked the re-request review, and github removed @GaryQian . @stuartmorgan maybe you can request review again. Thanks.

chunfengyao committed:
Fixed # remove the sign ":" before project's name.

GaryQian commented:
So in Gradle, the `:` indicates that something is a subproject of something else. `:blah` represents a root project, while `hello:blah` represents a project `hello` with a subproject `blah`. Since the code here is specifically a subproject of the root project, the `.name` property should not return a "path" style value with colons. Instead, subprojects of subprojects are to be found in the `.subprojects` property. Thus, we should not expect to find colons in `subproject.name`.

stuartmorgan commented:
Thanks for that explanation!

chunfengyao commented:
> @stuartmorgan So in Gradle, the `:` indicates that something is a subproject of something else. `:blah` represents a root project, while `hello:blah` represents a project `hello` with a subproject `blah`. Since the code here is specifically a subproject of the root project, the `.name` property should not return a "path" style value with colons. Instead, subprojects of subprojects are to be found in the `.subprojects` property. Thus, we should not expect to find colons in `subproject.name` and the colon parsing/replacement is not necessary.

It has been discussed at [the beginning of this PR](https://github.com/flutter/flutter/pull/94645#discussion_r846550381). 
But from the docs provided by Gradle, It seems that colons was used in gradle task's name to split project's name and action. (I'm not familiar with the detail of gradle task.)
And I don't remember whether the colon was before subproject's name acturally , or the first version of this PR uses the task's name as the plugin name?
But CI check passed when I removed the colons replacement.

Thanks for everyone who helped this PR!
